### PR TITLE
Change indexing and predication to address non-exact threading dimensions

### LINKED ISSUE #1102: Out-of-bounds access when parallelized by non-exact parallel types
Related issues and PRs: #1052, #1099, #1100. 

The same example as #1090 also has a problem with accesses to `T1`, which is allocated on SMEM.

```
  __shared__ float T1[4];
  if ((((((nvfuser_index_t)blockIdx.x) * ((nvfuser_index_t)blockDim.y)) + ((nvfuser_index_t)threadIdx.y)) < T0.size[0])) {
    T1[((nvfuser_index_t)threadIdx.y)]
      = T0[(((((nvfuser_index_t)blockIdx.x) * ((nvfuser_index_t)blockDim.y)) + ((nvfuser_index_t)threadIdx.y)) * T0.stride[0])]
      + (float) 1;
  }
  __barrier_sync(0);
  if ((((((nvfuser_index_t)blockIdx.x) * ((nvfuser_index_t)blockDim.z)) + ((nvfuser_index_t)threadIdx.z)) < T0.size[0])) {
    T2[(((((nvfuser_index_t)blockIdx.x) * ((nvfuser_index_t)blockDim.z)) + ((nvfuser_index_t)threadIdx.z)) * 1)]
      = T1[((nvfuser_index_t)threadIdx.z)]
      + (float) 1;
  }
```
The predicates and indices look good now with #1100, at least they look like what we originally thought they should be. However, the extent of `T1.axis(1)` is 4, and that's parallelized with `TIDy` and `TIDz`, each of which is 6 and 7, so this results in out-of-bounds accesses with the `T1` smem buffer. Unfortunately, compute-sanitizer doesn't seem to be able to catch it.

Two possible approaches to the issue. 

1. Expand `T1` allocation to something like `max(blockDim.y, blockDim.z)`. This would avoid the out-of-bounds accesses, but it doesn't seem quite right. Since smem is block local, not all threads would have right values in the buffer.

2. Keep the allocation size as is. Adds predicates such as `threadIdx.y < 4` at the `T1` store expr, and `threadIdx.z < 4` at the `T1` load expr. 

Option 2 would also need some change in index generation. For example, this

```
      = T0[(((((nvfuser_index_t)blockIdx.x) * ((nvfuser_index_t)blockDim.y)) + ((nvfuser_index_t)threadIdx.y)) * T0.stride[0])]
```

should be

```
      = T0[(((((nvfuser_index_t)blockIdx.x) * 4) + ((nvfuser_index_t)threadIdx.y)) * T0.stride[0])]
```

### PR DESCRIPTION
**Pull request by naoyam**
Fixes #1102 

This PR implements the second approach mentioned in #1102 For example, indexing and predicates are changed from:

```
      = T0[(((((nvfuser_index_t)blockIdx.x) * ((nvfuser_index_t)blockDim.y)) + ((nvfuser_index_t)threadIdx.y)) * T0.stride[0])]
```

to:

```
 = T0[(((((nvfuser_index_t)blockIdx.x) * 4) + ((nvfuser_index_t)threadIdx.y)) * T0.stride[0])]
```

The use of `blockDim.y` is replaced by the extent of the second axis of `T0`, which is `4` in this case. This change only matters when a parallel type is not exact (in this case `TIDy`). 

The indexing change only needed to change `getExtent` in index_compute.cpp. However, we also need to predicate `threadIdx` and `blockIdx` to be smaller than IterDomain extents. That's implemented as `ParallelizedDomainPredicate` in predicate_compute.h.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 1]
```diff
@@ -783,18 +783,12 @@ void IndexCompute::run() {
...
-  // which is always supposed to have the correct parallelization, so
-  // if the extent of id is mapped in the extent map, that should be
-  // always the right one.
+  // Pick from extent_map_ if available. Previously parallel
+  // dimensions were ued (e.g., blockDim.x), however, it would result
+  // in out-of-bounds errors when the extent of IterDomain is smaller
+  // than the threading dimension.
   if (extent_map_.find(id) != extent_map_.end()) {
     return extent_map_.at(id);
-  } else if (isParallelTypeThread(id->parallelType())) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp:779]: Should we do this if the parallel type is known to not be exact? Or at least known to be a compile time constant?

I assume if we did the split as: `tv->split(..., NamedScalar::getParallelDim(ParallelType::TIDx));` then we can't produce a non-exact parallel mapping.

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:779]: If the parallel type of `id` is exact, `extent_map_.at(id)` should be equal to `GpuLower::current()->parallelDimensionMap().get(id->parallelType())`, which should be also equal to  `NamedScalar::getParallelDim(id->parallelType())`. Is there any particular preference?

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:779]: > I assume if we did the split as: tv->split(..., NamedScalar::getParallelDim(ParallelType::TIDx)); then we can't produce a non-exact parallel mapping.

Right, `TIDx` would be exact and the extent of the inner domain would be `blockDim.x`, so `extent_map_` should have `blockDim.x` mapped to the inner domain.

**Pull request merged**